# LAB 1 - Flight Departured from NYC in 2013

### Task 1 Performing EDA

In [1]:
import pandas as pd
import sweetviz as sv
import numpy as np
from scipy.stats import ttest_ind

In [2]:
dataset = pd.read_csv("flights.csv")

In [3]:
dataset.head()

year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
0  2013      1    1     517.0             515        2.0     830.0   
1  2013      1    1     533.0             529        4.0     850.0   
2  2013      1    1     542.0             540        2.0     923.0   
3  2013      1    1     544.0             545       -1.0    1004.0   
4  2013      1    1     554.0             600       -6.0     812.0   

   sched_arr_time  arr_delay carrier  flight tailnum origin dest  air_time  \
0             819       11.0      UA    1545  N14228    EWR  IAH     227.0   
1             830       20.0      UA    1714  N24211    LGA  IAH     227.0   
2             850       33.0      AA    1141  N619AA    JFK  MIA     160.0   
3            1022      -18.0      B6     725  N804JB    JFK  BQN     183.0   
4             837      -25.0      DL     461  N668DN    LGA  ATL     116.0   

   distance  hour  minute             time_hour  
0      1400     5      15  2013-01-01T05:00:00Z  
1      1416     5      29  2013-01-01T05:00:00Z  
2      1089     5      40  2013-01-01T05:00:00Z  
3      1576     5      45  2013-01-01T05:00:00Z  
4       762     6       0  2013-01-01T06:00:00Z

In [4]:
dataset.tail()

year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
336771  2013      9   30       NaN            1455        NaN       NaN   
336772  2013      9   30       NaN            2200        NaN       NaN   
336773  2013      9   30       NaN            1210        NaN       NaN   
336774  2013      9   30       NaN            1159        NaN       NaN   
336775  2013      9   30       NaN             840        NaN       NaN   

        sched_arr_time  arr_delay carrier  flight tailnum origin dest  \
336771            1634        NaN      9E    3393     NaN    JFK  DCA   
336772            2312        NaN      9E    3525     NaN    LGA  SYR   
336773            1330        NaN      MQ    3461  N535MQ    LGA  BNA   
336774            1344        NaN      MQ    3572  N511MQ    LGA  CLE   
336775            1020        NaN      MQ    3531  N839MQ    LGA  RDU   

        air_time  distance  hour  minute             time_hour  
336771       NaN       213    14      55  2013-09-30T14:00:00Z  
336772       NaN       198    22       0  2013-09-30T22:00:00Z  
336773       NaN       764    12      10  2013-09-30T12:00:00Z  
336774       NaN       419    11      59  2013-09-30T11:00:00Z  
336775       NaN       431     8      40  2013-09-30T08:00:00Z

In [5]:
dataset.shape

(336776, 19)

In [6]:
dataset.describe()

year          month            day       dep_time  sched_dep_time  \
count  336776.0  336776.000000  336776.000000  328521.000000   336776.000000   
mean     2013.0       6.548510      15.710787    1349.109947     1344.254840   
std         0.0       3.414457       8.768607     488.281791      467.335756   
min      2013.0       1.000000       1.000000       1.000000      106.000000   
25%      2013.0       4.000000       8.000000     907.000000      906.000000   
50%      2013.0       7.000000      16.000000    1401.000000     1359.000000   
75%      2013.0      10.000000      23.000000    1744.000000     1729.000000   
max      2013.0      12.000000      31.000000    2400.000000     2359.000000   

           dep_delay       arr_time  sched_arr_time      arr_delay  \
count  328521.000000  328063.000000   336776.000000  327346.000000   
mean       12.639070    1502.054999     1536.380220       6.895377   
std        40.210061     533.264132      497.457142      44.633292   
min       -43.000000       1.000000        1.000000     -86.000000   
25%        -5.000000    1104.000000     1124.000000     -17.000000   
50%        -2.000000    1535.000000     1556.000000      -5.000000   
75%        11.000000    1940.000000     1945.000000      14.000000   
max      1301.000000    2400.000000     2359.000000    1272.000000   

              flight       air_time       distance           hour  \
count  336776.000000  327346.000000  336776.000000  336776.000000   
mean     1971.923620     150.686460    1039.912604      13.180247   
std      1632.471938      93.688305     733.233033       4.661316   
min         1.000000      20.000000      17.000000       1.000000   
25%       553.000000      82.000000     502.000000       9.000000   
50%      1496.000000     129.000000     872.000000      13.000000   
75%      3465.000000     192.000000    1389.000000      17.000000   
max      8500.000000     695.000000    4983.000000      23.000000   

              minute  
count  336776.000000  
mean       26.230100  
std        19.300846  
min         0.000000  
25%         8.000000  
50%        29.000000  
75%        44.000000  
max        59.000000

In [7]:
dataset.columns

Index(['year', 'month', 'day', 'dep_time', 'sched_dep_time', 'dep_delay',
       'arr_time', 'sched_arr_time', 'arr_delay', 'carrier', 'flight',
       'tailnum', 'origin', 'dest', 'air_time', 'distance', 'hour', 'minute',
       'time_hour'],
      dtype='object')

In [8]:
dataset.nunique()

year                 1
month               12
day                 31
dep_time          1318
sched_dep_time    1021
dep_delay          527
arr_time          1411
sched_arr_time    1163
arr_delay          577
carrier             16
flight            3844
tailnum           4043
origin               3
dest               105
air_time           509
distance           214
hour                20
minute              60
time_hour         6936
dtype: int64

In [9]:
dataset['carrier'].unique()

array(['UA', 'AA', 'B6', 'DL', 'EV', 'MQ', 'US', 'WN', 'VX', 'FL', 'AS',
       '9E', 'F9', 'HA', 'YV', 'OO'], dtype=object)

In [10]:
dataset['origin'].unique()

array(['EWR', 'LGA', 'JFK'], dtype=object)

In [11]:
dataset['dest'].unique()

array(['IAH', 'MIA', 'BQN', 'ATL', 'ORD', 'FLL', 'IAD', 'MCO', 'PBI',
       'TPA', 'LAX', 'SFO', 'DFW', 'BOS', 'LAS', 'MSP', 'DTW', 'RSW',
       'SJU', 'PHX', 'BWI', 'CLT', 'BUF', 'DEN', 'SNA', 'MSY', 'SLC',
       'XNA', 'MKE', 'SEA', 'ROC', 'SYR', 'SRQ', 'RDU', 'CMH', 'JAX',
       'CHS', 'MEM', 'PIT', 'SAN', 'DCA', 'CLE', 'STL', 'MYR', 'JAC',
       'MDW', 'HNL', 'BNA', 'AUS', 'BTV', 'PHL', 'STT', 'EGE', 'AVL',
       'PWM', 'IND', 'SAV', 'CAK', 'HOU', 'LGB', 'DAY', 'ALB', 'BDL',
       'MHT', 'MSN', 'GSO', 'CVG', 'BUR', 'RIC', 'GSP', 'GRR', 'MCI',
       'ORF', 'SAT', 'SDF', 'PDX', 'SJC', 'OMA', 'CRW', 'OAK', 'SMF',
       'TUL', 'TYS', 'OKC', 'PVD', 'DSM', 'PSE', 'BHM', 'CAE', 'HDN',
       'BZN', 'MTJ', 'EYW', 'PSP', 'ACK', 'BGR', 'ABQ', 'ILM', 'MVY',
       'SBN', 'LEX', 'CHO', 'TVC', 'ANC', 'LGA'], dtype=object)

In [12]:
dataset.isnull().sum()

year                 0
month                0
day                  0
dep_time          8255
sched_dep_time       0
dep_delay         8255
arr_time          8713
sched_arr_time       0
arr_delay         9430
carrier              0
flight               0
tailnum           2512
origin               0
dest                 0
air_time          9430
distance             0
hour                 0
minute               0
time_hour            0
dtype: int64

In [13]:
dataset.isnull()

year  month    day  dep_time  sched_dep_time  dep_delay  arr_time  \
0       False  False  False     False           False      False     False   
1       False  False  False     False           False      False     False   
2       False  False  False     False           False      False     False   
3       False  False  False     False           False      False     False   
4       False  False  False     False           False      False     False   
...       ...    ...    ...       ...             ...        ...       ...   
336771  False  False  False      True           False       True      True   
336772  False  False  False      True           False       True      True   
336773  False  False  False      True           False       True      True   
336774  False  False  False      True           False       True      True   
336775  False  False  False      True           False       True      True   

        sched_arr_time  arr_delay  carrier  flight  tailnum  origin   dest  \
0                False      False    False   False    False   False  False   
1                False      False    False   False    False   False  False   
2                False      False    False   False    False   False  False   
3                False      False    False   False    False   False  False   
4                False      False    False   False    False   False  False   
...                ...        ...      ...     ...      ...     ...    ...   
336771           False       True    False   False     True   False  False   
336772           False       True    False   False     True   False  False   
336773           False       True    False   False    False   False  False   
336774           False       True    False   False    False   False  False   
336775           False       True    False   False    False   False  False   

        air_time  distance   hour  minute  time_hour  
0          False     False  False   False      False  
1          False     False  False   False      False  
2          False     False  False   False      False  
3          False     False  False   False      False  
4          False     False  False   False      False  
...          ...       ...    ...     ...        ...  
336771      True     False  False   False      False  
336772      True     False  False   False      False  
336773      True     False  False   False      False  
336774      True     False  False   False      False  
336775      True     False  False   False      False  

[336776 rows x 19 columns]

In [14]:
dataset.dropna(axis=0,how ='any', inplace=True)
dataset.drop(axis=1,columns=['dep_time','sched_dep_time','sched_arr_time','arr_time','time_hour'],inplace=True)

In [15]:
dataset.isnull().sum()

year         0
month        0
day          0
dep_delay    0
arr_delay    0
carrier      0
flight       0
tailnum      0
origin       0
dest         0
air_time     0
distance     0
hour         0
minute       0
dtype: int64

### Task 2 Different Reports comparing data with each other

In [16]:
report = sv.analyze(dataset,target_feat= "arr_delay")
report.show_html()

                                             |          | [  0%]   00:00 -> (? left)

Report SWEETVIZ_REPORT.html was generated! NOTEBOOK/COLAB USERS: the web browser MAY not pop up, regardless, the report IS saved in your notebook/colab files.


### Task 3 Printing top 20,0000 data of UA and DL carrier

In [17]:
a = dataset[dataset['carrier'] == 'UA']
a = a.sort_values(by=["arr_delay"], ascending=False)
a = a.head(20000)
print(a)

        year  month  day  dep_delay  arr_delay carrier  flight tailnum origin  \
275124  2013      7   26      483.0      455.0      UA     372  N577UA    LGA   
182153  2013      4   18      427.0      435.0      UA     479  N811UA    LGA   
245329  2013      6   25      420.0      424.0      UA    1543  N76515    LGA   
89634   2013     12    7      374.0      422.0      UA     394  N842UA    EWR   
259405  2013      7   10      287.0      414.0      UA     745  N519UA    LGA   
...      ...    ...  ...        ...        ...     ...     ...     ...    ...   
7114    2013      1    9       -6.0        3.0      UA     601  N832UA    LGA   
212730  2013      5   21       19.0        3.0      UA     535  N557UA    JFK   
131131  2013      2   23        1.0        3.0      UA    1183  N77530    EWR   
298520  2013      8   20        2.0        3.0      UA     112  N26208    EWR   
203492  2013      5   11       21.0        3.0      UA    1412  N37293    EWR   

       dest  air_time  dist

In [18]:
b = dataset[dataset['carrier'] == 'DL']
b = b.sort_values(by=["arr_delay"], ascending=False)
b = b.head(20000)
print(b)

        year  month  day  dep_delay  arr_delay carrier  flight tailnum origin  \
173992  2013      4   10      960.0      931.0      DL    2391  N959DL    JFK   
151974  2013      3   17      911.0      915.0      DL    2119  N927DA    LGA   
270987  2013      7   22      898.0      895.0      DL    2047  N6716C    LGA   
95530   2013     12   14      825.0      856.0      DL    2391  N939DL    JFK   
247040  2013      6   27      899.0      850.0      DL    2007  N3762Y    JFK   
...      ...    ...  ...        ...        ...     ...     ...     ...    ...   
309962  2013      9    2       -2.0       -4.0      DL    1743  N3758Y    JFK   
35717   2013     10   10       -3.0       -4.0      DL     181  N337NW    LGA   
310051  2013      9    2       -8.0       -4.0      DL    2047  N642DL    LGA   
98416   2013     12   17       -1.0       -4.0      DL    1935  N333NB    LGA   
38683   2013     10   13        3.0       -4.0      DL    2459  N3737C    JFK   

       dest  air_time  dist

In [19]:
temp = pd.concat([a, b], axis = 0)
temp = temp.rename(columns = {'carrier' :'company_name', 'arr_delay':'delayed_arrival_flights'}, inplace = False)
temp = temp[["company_name","delayed_arrival_flights"]]
temp.to_csv(f'pop_data.csv', index=False)

### Task 4 Best 1,000 data of both carriers

In [20]:
pop_data = pd.read_csv("pop_data.csv")

In [21]:
temp = pop_data[pop_data['company_name'] == 'DL']
temp = temp.head(1000)
temp['sample_id'] = 1
temp.to_csv(f'dl.csv', index=False)
print(temp)

      company_name  delayed_arrival_flights  sample_id
20000           DL                    931.0          1
20001           DL                    915.0          1
20002           DL                    895.0          1
20003           DL                    856.0          1
20004           DL                    850.0          1
...            ...                      ...        ...
20995           DL                    129.0          1
20996           DL                    129.0          1
20997           DL                    129.0          1
20998           DL                    129.0          1
20999           DL                    129.0          1

[1000 rows x 3 columns]


In [22]:
temp = pop_data[pop_data['company_name'] == 'UA']
temp = temp.head(1000)
temp['sample_id'] = 2
temp.to_csv(f'ua.csv', index=False)
print(temp)

    company_name  delayed_arrival_flights  sample_id
0             UA                    455.0          2
1             UA                    435.0          2
2             UA                    424.0          2
3             UA                    422.0          2
4             UA                    414.0          2
..           ...                      ...        ...
995           UA                    141.0          2
996           UA                    141.0          2
997           UA                    141.0          2
998           UA                    141.0          2
999           UA                    141.0          2

[1000 rows x 3 columns]


### Task 5 Create two new sample data

In [23]:
dl = pd.read_csv("dl.csv")
ua = pd.read_csv("ua.csv")
temp = pd.concat([dl, ua], axis = 0)
temp.to_csv(f'samples.csv', index=False)

### Task 6 Calculate Standard Error and Mean

In [24]:
se_dl = np.std(dl['delayed_arrival_flights'], ddof=1) / np.sqrt(np.size(dl['delayed_arrival_flights']))
print(se_dl)

3.4595172894315587


In [25]:
mean_dl = dl['delayed_arrival_flights'].mean()
print(mean_dl)

214.628


In [26]:
se_ua = np.std(ua['delayed_arrival_flights'], ddof=1) / np.sqrt(np.size(ua['delayed_arrival_flights']))
print(se_ua)

1.939718458616792


In [27]:
mean_ua = ua['delayed_arrival_flights'].mean()
print(mean_ua)

207.553


### Task 7 Upper value and Lower Value

##### Mean delayed arrival of delta airlines = 214.63
##### Standard error for delta airlines = 3.46
##### Total number of delta airlines flights = 1000

##### Mean delayed arrival of united airlines = 207.55  
##### standard error for united airlines = 1.94
##### Total number of united airlines flights = 1000
##### z-score from the z table mentioned before = 1.96

In [28]:
z = 1.96
lcb_dl = mean_dl - z* se_dl  #lower limit of the CI for DA
ucb_dl = mean_dl + z* se_dl  #upper limit of the CI for DA
lcb_ua = mean_ua - z* se_ua  #lower limit of the CI for UA
ucb_ua = mean_ua + z* se_ua  #upper limit of the CI for UA
print("UA: " + str(round(lcb_ua, 2)) + " - " + str(round(mean_ua, 2)) + " - " + str(round(ucb_ua, 2)))
print("DL: " + str(round(lcb_dl, 2)) + " - " + str(round(mean_dl, 2)) + " - " + str(round(ucb_dl, 2)))

UA: 203.75 - 207.55 - 211.35
DL: 207.85 - 214.63 - 221.41


### Task 8 Applying T-Test Method

In [29]:
def compare_2_airlines(arr_1, arr_2, alpha, sample_size):
    stat, p = ttest_ind(arr_1, arr_2)
    print('Statistics=%.3f, p=%.3f' % (stat, p))
    if p > alpha:
        print('Delta airlines are delayed more than United airlines (fail to reject H0)')
    else:
        print('United airlines are delayed more than Delta airlines (reject H0)')

In [30]:
sample_size = 1000
compare_2_airlines(dl['delayed_arrival_flights'], ua['delayed_arrival_flights'], 0.05, sample_size)

Statistics=1.784, p=0.075
Delta airlines are delayed more than United airlines (fail to reject H0)


### This proves that Delta airlines are delayed more than United airlines using T-test